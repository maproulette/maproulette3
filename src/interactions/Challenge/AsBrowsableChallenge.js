import _isFinite from 'lodash/isFinite'

/**
 * AsBrowsableChallenge adds functionality to a Challenge related to browsing
 */
export class AsBrowsableChallenge {
  constructor(challenge) {
    Object.assign(this, challenge)
  }

  browseURL() {
    if (!_isFinite(this.id)) {
      return '/browse/challenges'
    }
    return `/browse/challenges/${this.id}`
  }
}

export default challenge => new AsBrowsableChallenge(challenge)

