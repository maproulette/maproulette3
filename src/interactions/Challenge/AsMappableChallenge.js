import _isFinite from 'lodash/isFinite'
import { basemapLayerSource }
       from '../../services/VisibleLayer/LayerSources'

/**
 * AsMappableChallenge adds functionality to a Challenge related to mapping.
 */
export class AsMappableChallenge {
  constructor(challenge) {
    Object.assign(this, challenge)
  }

  defaultLayerSource() {
    if (!_isFinite(this.id)) {
      return null
    }
    return basemapLayerSource(this.defaultBasemap,
                              this.defaultBasemapId)
  }
}

export default challenge => new AsMappableChallenge(challenge)
