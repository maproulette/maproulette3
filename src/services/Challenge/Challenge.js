import { normalize, schema } from 'normalizr'
import _get from 'lodash/get'
import _isObject from 'lodash/isObject'
import _compact from 'lodash/compact'
import _pick from 'lodash/pick'
import _map from 'lodash/map'
import _keys from 'lodash/keys'
import _values from 'lodash/values'
import _flatten from 'lodash/flatten'
import _clone from 'lodash/clone'
import _cloneDeep from 'lodash/cloneDeep'
import _isEmpty from 'lodash/isEmpty'
import _isString from 'lodash/isString'
import _isNumber from 'lodash/isNumber'
import _isArray from 'lodash/isArray'
import { defaultRoutes as api } from '../Server/Server'
import Endpoint from '../Server/Endpoint'
import RequestStatus from '../Server/RequestStatus'
import genericEntityReducer from '../Server/GenericEntityReducer'
import { commentSchema, receiveComments } from '../Comment/Comment'
import { projectSchema, fetchProject } from '../Project/Project'
import { logoutUser } from '../User/User'
import { toLatLngBounds } from '../MapBounds/MapBounds'
import { buildError,
         buildServerError,
         addError } from '../Error/Error'
import { RECEIVE_CHALLENGES,
         REMOVE_CHALLENGE } from './ChallengeActions'
import { parseQueryString } from '../Search/Search'

// normalizr schema
export const challengeSchema = function() {
  return new schema.Entity('challenges', {parent: projectSchema()})
}

/**
 * normalizr denormalization schema, which will also pull in comments (fetched
 * separately, so not used in normal schema)
 */
export const challengeDenormalizationSchema = function() {
  return new schema.Entity('challenges', {
    parent: projectSchema(),
    comments: [ commentSchema() ],
  })
}

// redux actions -- see Server/ChallengeActions

// redux action creators

/**
 * Add or update challenge data in the redux store
 */
export const receiveChallenges = function(normalizedEntities,
                                          status = RequestStatus.success) {
  return {
    type: RECEIVE_CHALLENGES,
    status,
    entities: normalizedEntities,
    receivedAt: Date.now()
  }
}

/**
 * Remove a challenge from the redux store
 */
export const removeChallenge = function(challengeId) {
  return {
    type: REMOVE_CHALLENGE,
    challengeId,
    receivedAt: Date.now()
  }
}

// async action creators

/**
 * Retrieve all featured challenges, up to the given limit.
 *
 * @param {number} limit
 */
export const fetchFeaturedChallenges = function(limit = 50) {
  return function(dispatch) {
    return new Endpoint(
      api.challenges.featured,
      {schema: [ challengeSchema() ], params: {limit}}
    ).execute().then(normalizedResults => {
      dispatch(receiveChallenges(normalizedResults.entities))
      return normalizedResults
    }).catch((error) => {
      dispatch(addError(buildError(
        "Challenge.fetchFailure", "Unable to retrieve latest challenge data from server."
      )))

      console.log(error.response || error)
    })
  }
}

/**
 * Fetches challenges that contain any of the given keywords/tags,
 * up to the given limit.
 *
 * @param {string} tags - comma-separated keywords/tags to include in search.
 * @param {number} limit
 */
export const fetchChallengesWithKeywords = function(keywords, limit=50) {
  return function(dispatch) {
    const keywordString = _isArray(keywords) ? keywords.join(',') : keywords

    return new Endpoint(
      api.challenges.withKeywords,
      {schema: [ challengeSchema() ], params: {ce: 'true', ct: keywordString, limit}}
    ).execute().then(normalizedResults => {
      dispatch(receiveChallenges(normalizedResults.entities))
      return normalizedResults
    }).catch((error) => {
      dispatch(addError(buildError(
        "Challenge.fetchFailure", "Unable to retrieve latest challenge data from server."
      )))

      console.log(error.response || error)
    })
  }
}

/**
 * Fetches a batch of enabled challenges, up to the given limit.
 */
export const fetchEnabledChallenges = function(limit) {
  return fetchChallengesWithKeywords('', limit)
}

/**
 * Search challenges by name using the given query string, which may also
 * contain hashtags for narrowing by keyword/tag. e.g. "#france museum"
 * would search for challenges with the "france" keyword that had "museum"
 * in the challenge name (as determined by the server).
 *
 * @param {string} queryString
 * @param {number} limit
 */
export const searchChallenges = function(queryString, limit=50) {
  return function(dispatch) {
    const queryParts = parseQueryString(queryString)

    // setup query parameters desired by server.
    // cs: query string
    // ct: keywords/tags (comma-separated string)
    const queryParams = {limit}
    if (queryParts.query.length > 0) {
      queryParams.cs = queryParts.query
    }

    if (queryParts.tags.length > 0) {
      queryParams.ct = queryParts.tags
    }

    return new Endpoint(
      api.challenges.search,
      {schema: [ challengeSchema() ], params: queryParams}
    ).execute().then(normalizedResults => {
      dispatch(receiveChallenges(normalizedResults.entities))
      return normalizedResults
    }).catch((error) => {
      dispatch(addError(buildError(
        "Challenge.searchFailure", "Unable to search challenges on server."
      )))

      console.log(error.response || error)
    })
  }
}

/**
 * Fetch challenges that have one or more tasks located within the given
 * bounding box.
 *
 * @param bounds - a LatLngBounds instance or an array of [west, south, east, north]
 *                 values.
 * @param {number} limit
 */
export const fetchChallengesWithinBoundingBox = function(bounds, limit=50) {
  const boundsObject = toLatLngBounds(bounds)

  return function(dispatch) {
    return new Endpoint(
      api.challenges.withinBounds,
      {
        schema: [ challengeSchema() ],
        params: {tbb: boundsObject.toBBoxString(), limit}
      }
    ).execute().then(normalizedResults => {
      dispatch(receiveChallenges(normalizedResults.entities))
      return normalizedResults
    }).catch((error) => {
      dispatch(addError(buildError(
        "Challenge.searchFailure", "Unable to search challenges on server."
      )))

      console.log(error.response || error)
    })
  }
}

/**
 * Fetch action metrics for the given challenge (or all challenges
 * if none is given).
 */
export const fetchChallengeActions = function(challengeId = null) {
  return function(dispatch) {
    const challengeActionsEndpoint = new Endpoint(
      _isNumber(challengeId) ? api.challenge.actions : api.challenges.actions,
      {schema: [ challengeSchema() ], variables: {id: challengeId}}
    )

    return challengeActionsEndpoint.execute().then(normalizedResults => {
      dispatch(receiveChallenges(normalizedResults.entities))
    }).catch((error) => {
      if (error.response && error.response.status === 401) {
        // If we get an unauthorized, we assume the user is not logged
        // in (or no longer logged in with the server). There's nothing to
        // do for this request except ensure we know the user is logged out.
        dispatch(logoutUser())
      }
      else {
        dispatch(addError(buildError(
          "Challenge.fetchFailure", "Unable to retrieve latest challenge data from server."
        )))

        console.log(error.response || error)
      }
    })
  }
}

/**
 * Fetch activity timeline for the given challenge
 */
export const fetchChallengeActivity = function(challengeId, startDate, endDate) {
  return function(dispatch) {
    const params = {}
    if (startDate) {
      params.start = startDate.toISOString()
    }

    if (endDate) {
      params.end = endDate.toISOString()
    }

    return new Endpoint(
      api.challenge.activity, {
        variables: {id: challengeId},
        params,
      }
    ).execute().then(rawActivity => {
      const normalizedResults = {
        entities: {
          challenges: {
            [challengeId]: {activity: rawActivity},
          }
        }
      }

      return dispatch(receiveChallenges(normalizedResults.entities))
    }).catch((error) => {
      if (error.response && error.response.status === 401) {
        // If we get an unauthorized, we assume the user is not logged
        // in (or no longer logged in with the server). There's nothing to
        // do for this request except ensure we know the user is logged out.
        dispatch(logoutUser())
      }
      else {
        dispatch(addError(buildError(
          "Challenge.fetchFailure", "Unable to retrieve latest challenge data from server."
        )))

        console.log(error.response || error)
      }
    })
  }
}

/**
 * Fetch comments for the given challenge
 */
export const fetchChallengeComments = function(challengeId) {
  return function(dispatch) {
    return new Endpoint(
      api.challenge.comments, {variables: {id: challengeId}}
    ).execute().then(rawComments => {
      // Comments are indexed by task id, so we need to extract them
      // into a single array and manually denormalize
      const commentArray = _flatten(_values(rawComments))
      const normalizedComments = normalize(commentArray, [ commentSchema() ])
      dispatch(receiveComments(normalizedComments.entities))

      if (_isObject(normalizedComments.entities.comments)) {
        // Inject comment ids into challenge.
        dispatch(receiveChallenges({
          challenges: {
            [challengeId]: {
              id: challengeId,
              comments: _map(_keys(normalizedComments.entities.comments),
                             id => parseInt(id, 10)),
            }
          }
        }))
      }

      return normalizedComments
    })
  }
}

/**
 * Fetch child challenges belonging to the given project, up to the given
 * limit.
 */
export const fetchProjectChallenges = function(projectId, limit=50) {
  return function(dispatch) {
    return new Endpoint(
      api.project.challenges,
      {
        schema: [ challengeSchema() ],
        variables: {id: projectId},
        params: {limit}
      }
    ).execute().then(normalizedResults => {
      dispatch(receiveChallenges(normalizedResults.entities))
      return normalizedResults
    }).catch((error) => {
      dispatch(addError(buildError(
        "Challenge.fetchFailure",
        "Unable to retrieve latest challenge data from server."
      )))

      console.log(error.response || error)
    })
  }
}

/**
 * Fetch data for the given challenge. Normally that data will be added to the
 * redux store, but that can be suppressed with the supressReceive flag.
 */
export const fetchChallenge = function(challengeId, suppressReceive = false) {
  return function(dispatch) {
    return new Endpoint(
      api.challenge.single,
      {schema: challengeSchema(), variables: {id: challengeId}}
    ).execute().then(normalizedResults => {
      if (!suppressReceive) {
        dispatch(receiveChallenges(normalizedResults.entities))
      }

      return normalizedResults
    }).catch((error) => {
      dispatch(addError(buildError(
        "Challenge.fetchFailure", "Unable to retrieve latest challenge data from server."
      )))

      console.log(error.response || error)
    })
  }
}

/**
 * Convenience function that fetches the given challenge plus its parent project,
 * so that a fully denormalized challenge entity can be reliably created.
 */
export const loadCompleteChallenge = function(challengeId) {
  return function(dispatch) {
    if (!challengeId) {
      return null
    }

    return dispatch(fetchChallenge(challengeId, true)).then(normalizedChallengeResults =>
      fetchParentProject(dispatch, normalizedChallengeResults).then(() =>
        dispatch(receiveChallenges(normalizedChallengeResults.entities))
      ).then(() => normalizedChallengeResults)
    ).catch((error) => {
      dispatch(addError(buildError(
        "Challenge.fetchFailure", "Unable to retrieve latest challenge data from server."
      )))

      console.log(error.response || error)
    })
  }
}

/**
 * Saves the given challenge (either creating it or updating it, depending on
 * whether it already has an id) and updates the redux store with the latest
 * version from the server.
 */
export const saveChallenge = function(originalChallengeData) {
  return function(dispatch) {
    // The server wants keywords/tags represented as a comma-separated string.
    let challengeData = _clone(originalChallengeData)
    if (_isArray(challengeData.tags)) {
      challengeData.tags = challengeData.tags.join(',')
    }

    // If there is local GeoJSON content being transmitted as a string, parse
    // it into JSON first.
    if (_isString(challengeData.localGeoJSON) &&
        !_isEmpty(challengeData.localGeoJSON)) {
      challengeData.localGeoJSON = JSON.parse(challengeData.localGeoJSON)
    }

    // We need to remove any old challenge keywords first, prior to the
    // update.
    return removeChallengeKeywords(challengeData.id, challengeData.removedTags).then(() => {
      challengeData = _pick(challengeData, // fields in alphabetical order
        ['blurb', 'challengeType', 'checkinComment', 'customBasemap',
        'defaultBasemap', 'defaultPriority', 'defaultZoom', 'description',
        'difficulty', 'enabled', 'featured', 'highPriorityRule', 'id',
        'instruction', 'localGeoJSON', 'lowPriorityRule', 'maxZoom',
        'mediumPriorityRule', 'minZoom', 'name', 'overpassQL', 'remoteGeoJson',
        'status', 'tags', 'updateTasks'])

      // Setup the save function to either edit or create the challenge
      // depending on whether it has an id.
      const saveEndpoint = new Endpoint(
        _isNumber(challengeData.id) ? api.challenge.edit : api.challenge.create,
        {
          schema: challengeSchema(),
          variables: {id: challengeData.id},
          json: challengeData
        }
      )

      return saveEndpoint.execute().then(normalizedResults => {
        dispatch(receiveChallenges(normalizedResults.entities))
        return _get(normalizedResults, `entities.challenges.${normalizedResults.result}`)
      }).catch((error) => {
        if (error.response && error.response.status === 401) {
          // If we get an unauthorized, we assume the user is not logged
          // in (or no longer logged in with the server).
          dispatch(logoutUser())
          dispatch(addError(buildError(
            "User.unauthorized", "Please sign in to continue."
          )))
        }
        else {
          console.log(error.response || error)
          buildServerError(
            "Challenge.saveFailure", "Unable to save your changes", error
          ).then(errorObject => dispatch(addError(errorObject)))
        }
      })
    })
  }
}

/**
 * Deletes the given challenge from the server.
 */
export const deleteChallenge = function(challengeId) {
  return function(dispatch) {
    return new Endpoint(
      api.challenge.delete,
      {variables: {id: challengeId}}
    ).execute().then(() =>
      dispatch(removeChallenge(challengeId))
    ).catch((error) => {
      if (error.response && error.response.status === 401) {
        // If we get an unauthorized, we assume the user is not logged
        // in (or no longer logged in with the server).
        dispatch(logoutUser())
        dispatch(addError(buildError(
          "User.unauthorized", "Please sign in to continue."
        )))
      }
      else {
        dispatch(addError(buildError(
          "Challenge.deleteFailure", "Unable to delete challenge."
        )))

        console.log(error.response || error)
      }
    })
  }
}

/**
 * Retrieves the parent project embedded in the given normalized challenge
 * results.
 *
 * > Note that if the results contain multiple challenges, only the
 * > parent project of the first result is retrieved.
 */
const fetchParentProject = function(dispatch, normalizedChallengeResults) {
  const challengeId = _isArray(normalizedChallengeResults.result) ?
                      normalizedChallengeResults.result[0] :
                      normalizedChallengeResults.result
  const projectId = normalizedChallengeResults.entities.challenges[challengeId].parent
  return dispatch(fetchProject(projectId))
}

/**
 * Removes the given oldKeywords from the given challenge. If no
 * keywords are provided, nothing is done.
 *
 * @private
 *
 * @returns a Promise
 */
const removeChallengeKeywords = function(challengeId, oldKeywords=[]) {
  // strip empty tags
  const toRemove =
    _compact(_map(oldKeywords, tag => _isEmpty(tag) ? null : tag))

  // if no keywords given, nothing to do.
  if (toRemove.length === 0) {
    return Promise.resolve()
  }
  else {
    return new Endpoint(
      api.challenge.removeKeywords, {
        variables: {id: challengeId},
        params: {tags: toRemove.join(',')},
      }
    ).execute()
  }
}

/**
 * reduceChallengesFurther will be invoked by the genericEntityReducer function to
 * perform additional reduction on challenge entities.
 *
 * @private
 */
const reduceChallengesFurther = function(mergedState, oldState, challengeEntities) {
  // The generic reduction will merge arrays, creating a union of values.
  // We don't want that for keywords/tags: we want to replace the old array
  // with the new one.
  challengeEntities.forEach(entity => {
    if (_isArray(entity.tags)) {
      mergedState[entity.id].tags = entity.tags
    }
  })
}

// redux reducers
export const challengeEntities = function(state, action) {
  if (action.type === REMOVE_CHALLENGE) {
    const mergedState = _cloneDeep(state)
    delete mergedState[action.challengeId]
    return mergedState
  }
  else {
    return genericEntityReducer(
      RECEIVE_CHALLENGES, 'challenges', reduceChallengesFurther)(state, action)
  }
}
