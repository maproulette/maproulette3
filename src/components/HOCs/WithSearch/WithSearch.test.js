import React, { Component } from 'react'
import { mapStateToProps, mapDispatchToProps,
         _WithSearch } from './WithSearch'
import { setSort, removeSort,
         setFilters, removeFilters,
         setSearch, clearSearch,
         performSearch } from '../../../services/Search/Search'
import { extendedFind } from '../../../services/Challenge/Challenge'
import _isFunction from 'lodash/isFunction'
import _cloneDeep from 'lodash/cloneDeep'

jest.mock('../../../services/Search/Search')
jest.mock('../../../services/Challenge/Challenge')

let basicState = null
let WrappedComponent = null
let WrappedChallengesComponent = null

let SEARCH_NAME = 'challenges'

beforeEach(() => {
  basicState = {
    searchGroup: 'challenges',
    currentSearch: {
      challenges: {
        sort: {
          sortBy: 'created',
          direction: 'desc'
        },
        filters: {
          difficulty: 'hard',
          featured: true,
          keywords: 'testing',
          location: 'USA',
        },
        mapBounds: {
          bounds: [1,2,3,4]
        }
      },
      searchQueryTest: "searchQueryValue",
    },
    setChallengeSearchMapBounds: jest.fn(),
    performSearch: jest.fn(),
  }

  WrappedComponent = _WithSearch(
    () => <div className="child" />,
    "searchQueryTest",
    jest.fn()
  )

  WrappedChallengesComponent = _WithSearch(
    () => <div className="child" />,
    "challenges",
    jest.fn()
  )
})

test("mapStateToProps maps 'difficulty', 'featured', 'keywords', 'location' to searchFilters", () => {
  const mappedProps = mapStateToProps(basicState, 'challenges')

  expect(
    mappedProps.searchFilters
  ).toMatchObject(basicState.currentSearch.challenges.filters)
})

test("mapStateToProps maps 'sortBy', 'direction', to searchSort", () => {
  const mappedProps = mapStateToProps(basicState, 'challenges')

  expect(
    mappedProps.searchSort
  ).toMatchObject(basicState.currentSearch.challenges.sort)
})

test("mapDispatchToProps maps call setSearchFilters", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch, basicState, 'challenges')
  const filterSetting = {difficulty: 1}

  mappedProps.setSearchFilters(filterSetting)
  expect(dispatch).toBeCalled()
  expect(setFilters).toBeCalledWith('challenges', filterSetting)
})

test("mapDispatchToProps maps call removeSearchFilters", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch, basicState, 'challenges')
  const filterSetting = {a: 1}

  mappedProps.removeSearchFilters(filterSetting)
  expect(dispatch).toBeCalled()
  expect(removeFilters).toBeCalledWith('challenges', filterSetting)
})

test("mapDispatchToProps maps call setKeywordFilter", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch, basicState, 'challenges')
  const keywords = ["foo", "bar"]

  mappedProps.setKeywordFilter(keywords)
  expect(dispatch).toBeCalled()
  expect(setFilters).toBeCalledWith('challenges',
                                    expect.objectContaining({keywords}))
})

test("mapDispatchToProps maps call setSearchSort: 'name' sort gets an 'asc' direction", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch, basicState, 'challenges')
  const sortSetting = {sortBy: 'name'}

  mappedProps.setSearchSort(sortSetting)
  expect(dispatch).toBeCalled()
  expect(setSort).toBeCalledWith('challenges', {sortBy: 'name', direction: 'asc'})
})

test("mapDispatchToProps maps call setSearchSort: 'created' sort gets an 'desc' direction", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch, basicState, 'challenges')
  const sortSetting = {sortBy: 'created'}

  mappedProps.setSearchSort(sortSetting)
  expect(dispatch).toBeCalled()
  expect(setSort).toBeCalledWith('challenges', {sortBy: 'created', direction: 'desc'})
})

test("mapDispatchToProps maps call setSearchSort: 'xxx' sort is set to null", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch, basicState, 'challenges')
  const sortSetting = {sortBy: 'xxx'}

  mappedProps.setSearchSort(sortSetting)
  expect(dispatch).toBeCalled()
  expect(setSort).toBeCalledWith('challenges', {sortBy: null, direction: null})
})

test("mapDispatchToProps maps call removeSearchFilters", () => {
  const dispatch = jest.fn()
  const mappedProps = mapDispatchToProps(dispatch, basicState, 'challenges')
  const sortSetting = {a: 1}

  mappedProps.removeSearchSort(sortSetting)
  expect(dispatch).toBeCalled()
  expect(removeSort).toBeCalledWith('challenges', sortSetting)
})

test("searchQueries are passed through to the wrapped component", () => {
  const wrapper = shallow(
    <WrappedComponent {...basicState} />
  )

  expect(
    wrapper.props().searchQuery
  ).toBe(basicState.currentSearch.searchQueryTest)

  expect(wrapper).toMatchSnapshot()
})

test("mapStateToProps maps currentSearch", () => {
  const mappedProps = mapStateToProps(basicState)
  expect(mappedProps.currentSearch).toEqual(basicState.currentSearch)

  expect(mappedProps).toMatchSnapshot()
})

test("mapDispatchToProps maps function setSearch", () => {
  const dispatch = jest.fn(() => Promise.resolve())
  const mappedProps = mapDispatchToProps(dispatch, {}, 'challenges')

  mappedProps.setSearch("query", "searchGroup")
  expect(dispatch).toBeCalled()
  expect(setSearch).toBeCalledWith("searchGroup", "query")

  expect(mappedProps).toMatchSnapshot()
})

test("mapDispatchToProps maps function clearSearch", () => {
  const dispatch = jest.fn(() => Promise.resolve())
  const mappedProps = mapDispatchToProps(dispatch, {}, 'challenges')

  mappedProps.clearSearch("searchGroup")
  expect(dispatch).toBeCalled()
  expect(clearSearch).toBeCalledWith("searchGroup")
})

test("searchQueries.searchGroup is passed through to the wrapped component", () => {
  const wrapper = shallow(
    <WrappedComponent {...basicState} />
  )

  expect(
    wrapper.props().searchGroup
  ).toBe(basicState.searchGroup)

  expect(wrapper).toMatchSnapshot()
})

test("mapDispatchToProps maps function performSearch", () => {
  const dispatch = jest.fn(() => Promise.resolve())
  const mappedProps = mapDispatchToProps(dispatch, {}, "searchGroup")
  const someProps = {foo: "foo"}

  mappedProps.performSearch("query", "searchProjects", someProps)
  expect(dispatch).toBeCalled()
  expect(performSearch).toBeCalledWith("searchGroup", "query", "searchProjects", someProps)
  expect(mappedProps).toMatchSnapshot()
})

test("moving the map doesn't signal challenges updates if not filtering on map bounds", () => {
  const wrapper = shallow(
    <WrappedChallengesComponent {...basicState} />
  )


  const newState = _cloneDeep(basicState)
  newState.currentSearch.challenges.mapBounds = {bounds: [0,0,0,1]}

  wrapper.instance().componentDidUpdate(newState)

  expect(basicState.performSearch).not.toBeCalled()
})

test("moving the map does perform new search if filtering within map bounds", () => {
  basicState.searchFilters = {location: 'withinMapBounds'}
  const wrapper = shallow(
    <WrappedChallengesComponent {...basicState} />
  )

  const newState = _cloneDeep(basicState)
  newState.currentSearch.challenges.mapBounds = {bounds: [0,0,0,1]}

  const didNewSearch = wrapper.instance().componentDidUpdate(newState)

  //expect(didNewSearch).toBe(true)
})

test("changing filters performs new search", () => {
  const wrapper = shallow(
    <WrappedChallengesComponent {...basicState} />
  )

  const newState = {currentSearch: {challenges: {filters: {difficulty: 'easy'}}}}
  const didNewSearch = wrapper.instance().componentDidUpdate(newState)

  //expect(didNewSearch).toBe(true)
})
