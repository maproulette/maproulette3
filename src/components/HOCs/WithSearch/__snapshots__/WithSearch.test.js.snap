// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapDispatchToProps maps function performSearch 1`] = `
Object {
  "clearMapBounds": [Function],
  "clearSearch": [Function],
  "clearSearchFilters": [Function],
  "locateMapToUser": [Function],
  "performSearch": [Function],
  "removeSearchFilters": [Function],
  "removeSearchSort": [Function],
  "setChallengeBrowseMapBounds": [Function],
  "setChallengeOwnerMapBounds": [Function],
  "setChallengeSearchMapBounds": [Function],
  "setKeywordFilter": [Function],
  "setSearch": [Function],
  "setSearchFilters": [Function],
  "setSearchPage": [Function],
  "setSearchSort": [Function],
  "setTaskMapBounds": [Function],
}
`;

exports[`mapDispatchToProps maps function setSearch 1`] = `
Object {
  "clearMapBounds": [Function],
  "clearSearch": [Function],
  "clearSearchFilters": [Function],
  "locateMapToUser": [Function],
  "performSearch": [Function],
  "removeSearchFilters": [Function],
  "removeSearchSort": [Function],
  "setChallengeBrowseMapBounds": [Function],
  "setChallengeOwnerMapBounds": [Function],
  "setChallengeSearchMapBounds": [Function],
  "setKeywordFilter": [Function],
  "setSearch": [Function],
  "setSearchFilters": [Function],
  "setSearchPage": [Function],
  "setSearchSort": [Function],
  "setTaskMapBounds": [Function],
}
`;

exports[`mapStateToProps maps currentSearch 1`] = `
Object {
  "currentSearch": Object {
    "challenges": Object {
      "filters": Object {
        "difficulty": "hard",
        "featured": true,
        "keywords": "testing",
        "location": "USA",
      },
      "mapBounds": Object {
        "bounds": Array [
          1,
          2,
          3,
          4,
        ],
      },
      "sort": Object {
        "direction": "desc",
        "sortBy": "created",
      },
    },
    "searchQueryTest": "searchQueryValue",
  },
  "mapBounds": Object {
    "bounds": Object {
      "_northEast": Object {
        "lat": 22.51255695405145,
        "lng": 96.15234375,
      },
      "_southWest": Object {
        "lat": -22.512556954051437,
        "lng": -6.152343750000001,
      },
    },
  },
  "searchCriteria": undefined,
  "searchFilters": Object {},
  "searchPage": Object {},
  "searchSort": Object {},
}
`;

exports[`searchQueries are passed through to the wrapped component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Component
    currentSearch={
      Object {
        "challenges": Object {
          "filters": Object {
            "difficulty": "hard",
            "featured": true,
            "keywords": "testing",
            "location": "USA",
          },
          "mapBounds": Object {
            "bounds": Array [
              1,
              2,
              3,
              4,
            ],
          },
          "sort": Object {
            "direction": "desc",
            "sortBy": "created",
          },
        },
        "searchQueryTest": "searchQueryValue",
      }
    }
    performSearch={[MockFunction]}
    searchGroup="challenges"
    setChallengeSearchMapBounds={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "clearSearch": [Function],
      "currentSearch": Object {
        "challenges": Object {
          "filters": Object {
            "difficulty": "hard",
            "featured": true,
            "keywords": "testing",
            "location": "USA",
          },
          "mapBounds": Object {
            "bounds": Array [
              1,
              2,
              3,
              4,
            ],
          },
          "sort": Object {
            "direction": "desc",
            "sortBy": "created",
          },
        },
        "searchQueryTest": "searchQueryValue",
      },
      "isLoading": false,
      "searchFunction": [MockFunction],
      "searchGroup": "challenges",
      "searchQueries": Object {
        "searchQueryTest": Object {
          "clearSearch": [Function],
          "searchQuery": "searchQueryValue",
          "setSearch": [Function],
        },
      },
      "searchQuery": "searchQueryValue",
      "setChallengeSearchMapBounds": [MockFunction],
      "setSearch": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "clearSearch": [Function],
        "currentSearch": Object {
          "challenges": Object {
            "filters": Object {
              "difficulty": "hard",
              "featured": true,
              "keywords": "testing",
              "location": "USA",
            },
            "mapBounds": Object {
              "bounds": Array [
                1,
                2,
                3,
                4,
              ],
            },
            "sort": Object {
              "direction": "desc",
              "sortBy": "created",
            },
          },
          "searchQueryTest": "searchQueryValue",
        },
        "isLoading": false,
        "searchFunction": [MockFunction],
        "searchGroup": "challenges",
        "searchQueries": Object {
          "searchQueryTest": Object {
            "clearSearch": [Function],
            "searchQuery": "searchQueryValue",
            "setSearch": [Function],
          },
        },
        "searchQuery": "searchQueryValue",
        "setChallengeSearchMapBounds": [MockFunction],
        "setSearch": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`searchQueries.searchGroup is passed through to the wrapped component 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Component
    currentSearch={
      Object {
        "challenges": Object {
          "filters": Object {
            "difficulty": "hard",
            "featured": true,
            "keywords": "testing",
            "location": "USA",
          },
          "mapBounds": Object {
            "bounds": Array [
              1,
              2,
              3,
              4,
            ],
          },
          "sort": Object {
            "direction": "desc",
            "sortBy": "created",
          },
        },
        "searchQueryTest": "searchQueryValue",
      }
    }
    performSearch={[MockFunction]}
    searchGroup="challenges"
    setChallengeSearchMapBounds={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "clearSearch": [Function],
      "currentSearch": Object {
        "challenges": Object {
          "filters": Object {
            "difficulty": "hard",
            "featured": true,
            "keywords": "testing",
            "location": "USA",
          },
          "mapBounds": Object {
            "bounds": Array [
              1,
              2,
              3,
              4,
            ],
          },
          "sort": Object {
            "direction": "desc",
            "sortBy": "created",
          },
        },
        "searchQueryTest": "searchQueryValue",
      },
      "isLoading": false,
      "searchFunction": [MockFunction],
      "searchGroup": "challenges",
      "searchQueries": Object {
        "searchQueryTest": Object {
          "clearSearch": [Function],
          "searchQuery": "searchQueryValue",
          "setSearch": [Function],
        },
      },
      "searchQuery": "searchQueryValue",
      "setChallengeSearchMapBounds": [MockFunction],
      "setSearch": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "clearSearch": [Function],
        "currentSearch": Object {
          "challenges": Object {
            "filters": Object {
              "difficulty": "hard",
              "featured": true,
              "keywords": "testing",
              "location": "USA",
            },
            "mapBounds": Object {
              "bounds": Array [
                1,
                2,
                3,
                4,
              ],
            },
            "sort": Object {
              "direction": "desc",
              "sortBy": "created",
            },
          },
          "searchQueryTest": "searchQueryValue",
        },
        "isLoading": false,
        "searchFunction": [MockFunction],
        "searchGroup": "challenges",
        "searchQueries": Object {
          "searchQueryTest": Object {
            "clearSearch": [Function],
            "searchQuery": "searchQueryValue",
            "setSearch": [Function],
          },
        },
        "searchQuery": "searchQueryValue",
        "setChallengeSearchMapBounds": [MockFunction],
        "setSearch": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
